### PR TITLE
fix(io): Improve readDelims() performance

### LINKED ISSUE #803: bufio.readLines is pathologically slow for very long lines
I found that bufio.readLines becomes pathologically slow when given large lines, such as a single 10MB line. Deno took 58 seconds to read and write this line, compared to under 0.1 seconds for Go, Python and Perl. Those other languages are able to read and write a 100MB line in under 0.3 seconds on my laptop, whereas I was not patient enough to wait for Deno bufio to finish that task.

I'm guessing that bufio is reallocing its buffer in small steps, which results in quadratic complexity, rather than the normal method which is to double the capacity of the buffer each time it needs to be expanded, which gives linear complexity. I can have a go at fixing this, but would like to get your go ahead first.

## Create a file containing a single 10MB line. I used Perl:

```
perl -e 'for (1..10000000) { print "X"; } print "\n";' > bigline-10M.txt
```

## readLine.ts: TypeScript Deno code to read and write the line

```
import { readLines } from "https://deno.land/std/io/bufio.ts";

const lines = readLines(Deno.stdin);

for await (const l of lines) {
        console.log(l);
        break;
}
```

## readLine.py: Python code to read and write the line

```
import sys
print(sys.stdin.readline(), end='')
```

## readLine.pl: Perl code to read and write the line

```
print scalar <>;
```

## readLine.go: Go code to read and write the line

```
package main

import "bufio"
import "os"
import "fmt"
import "log"

func main() {
        var buf []byte = make([]byte, 64*1024)
        scanner := bufio.NewScanner(os.Stdin)
        scanner.Buffer(buf, 1e12)
        if scanner.Scan() {
                fmt.Println(scanner.Text())
        }
        if err := scanner.Err(); err != nil {
                log.Fatal(err)
        }
}
```

## test.sh: script to time execution for each language

```
time() { command time -f '%e %U %S %C' "$@"; }
time deno run readLine.ts < bigline-10M.txt > /dev/null
time python readLine.py < bigline-10M.txt > /dev/null
time perl readLine.pl < bigline-10M.txt > /dev/null
time go build readLine.go
time ./readLine < bigline-10M.txt > /dev/null
```

## results

```
58.67 56.13 12.43 deno run readLine.ts
0.05 0.03 0.02 python readLine.py
0.01 0.00 0.00 perl readLine.pl
0.07 0.07 0.04 go build readLine.go
0.02 0.00 0.02 ./readLine
```

### PR DESCRIPTION
**Pull request by keroxp**
- fix #803
- After short research, I found that problem comes from `Buffer` itself.
- Stopped using `Buffer` and introduced `BytesList` that manage multiple Uint8Array like one single Uint8Array.
- This change has significant performance improvements against main branch especially for a big file.

## Improvements

v2 (de049d6)

| Size | main (sec) | this PR (sec) |
| --- | --- | --- |
| 1kb | 0.001 | 0.0006 |
| 10kb | 0.002 | 0.002 |
| 100kb | 0.008 | 0.005 |
| 1mb | 0.467 | 0.062 |
|10mb | 46.962 | 0.330 |
| 100mb | N/A (over 45min) | 3.407 |

v1 (59b3c76)

| Size | main (sec) | this PR (sec) |
| --- | --- | --- |
| 1kb | 0.001 | 0.002 |
| 10kb | 0.002 | 0.004 |
| 100kb | 0.008 | 0.01 |
| 1mb | 0.467 | 0.1 |
|10mb | 46.962 | 1.295 |
| 100mb | N/A (over 45min) | 15.791 |

### CODE REVIEW

File: bytes/bytes_list_test.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    bytes.equals(
+      arr.concat(),
+      new Uint8Array([
+        2,
+      ]),
+    ),
+    true,
+  );
+});
+Deno.test("[bytes] BytesList.shift 3", () => {
```
Comment (ry) [inline bytes/bytes_list_test.ts:131]: shift 100?

Comment (keroxp) [reply bytes/bytes_list_test.ts:131]: It's simply ordinal number.

**Comment by keroxp** [comment]
@ry Got more improvements by using iterator.

**Review by ry — APPROVED** [review]
LGTM

**Pull request merged**